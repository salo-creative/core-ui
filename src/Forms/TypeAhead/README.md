# TypeAhead

This is an input field that shows a bunch of suggested values to the user as they begin typing. These are displayed in a dropdown, which can be navigated with the arrow keys and submitted by pressing the return key.

The possible suggestions should be included by passing an array to the component in the suggestions prop.

If you pass a 'debounced' boolean to the TypeAhead, it will run the onChange function you pass as a debounced function 500ms after the last text input. This is designed to cut down on unnecessary asynchronous requests. Debounced is false by default, and if it's false, onChange will run every time instantaneously as the input changes.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import TypeAhead from '@salo/core-ui/Forms/TypeAhead';
```

Implement as follows

```javascript
<TypeAhead
    onChange={ (value) => {
        const newSuggestions = // Filter array of options
        setState({ suggestions: newSuggestions });
    } }
    label={ label }
    loading={ loading }
    helperText={ helperText }
    error={ error }
    required={ required }
    disabled={ disabled }
    suggestions={ state.suggestions }
    name='typeAhead'
/>
```

Asynchronous suggestions:

```javascript
<TypeAhead
    onChange={ (value) => {
        store.set({ value, skip: false });
    } }
    onSelect={ (value) => console.log('Selected item:', value) }
    label={ label }
    loading={ loading }
    helperText={ helperText }
    error={ error }
    required={ required }
    disabled={ disabled }
    suggestions={ get(data, 'user_list_permission_groups', []).map(group => {
        return { id: group.id, label: group.name };
    }) }
    value={ store.value }
    name='story'
    debounced
/>
```