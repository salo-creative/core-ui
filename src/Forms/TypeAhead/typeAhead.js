import React from 'react';
import PropTypes from 'prop-types';
import { debounce } from 'lodash';
import Input from '../Input';

// Components & Styles
import SuggestionsList from './typeAhead.suggestions';

// HELPERS & CONSTANTS
import { colours } from '../../helpers/colours';

const TypeAhead = (props) => {
  const {
    background,
    debounced,
    error,
    errorMessage,
    loading,
    onChange,
    onSelect,
    name,
    retryAction,
    suggestions,
    ...inputProps
  } = props;

  const [dropState, setDropState] = React.useState({ element: null, dropOpen: false });

  const [state, setState] = React.useState({
    activeSuggestion: 0,
    showSuggestions: false,
    userInput: ''
  });

  const onChangeDebounced = React.useCallback(debounce(onChange, 500), []);

  // Handle keyboard navigation of the drop element
  const navigate = (e) => {
    // Handle enter key
    if (e.keyCode === 13) {
      setState({
        ...state,
        activeSuggestion: 0,
        showSuggestions: false,
        userInput: suggestions[state.activeSuggestion].label
      });
      onSelect(suggestions[state.activeSuggestion]);
    }
    // Handle up arrow
    if (e.keyCode === 38) {
      if (state.activeSuggestion === 0) return;
      setState({ ...state, activeSuggestion: state.activeSuggestion - 1 });
    }
    // Handle down arrow
    if (e.keyCode === 40) {
      if (state.activeSuggestion - 1 === suggestions.length) return;
      setState({ ...state, activeSuggestion: state.activeSuggestion + 1 });
    }
  };

  // useEffect handler for navigate(), fires on state change
  React.useEffect(() => {
    window.addEventListener('keydown', navigate);
    return () => window.removeEventListener('keydown', navigate);
  }, [state]);

  const openDrop = (e) => {
    setDropState({ element: e.target, dropOpen: true });
  };

  // Handle when the user alters the Input
  const handleChange = (event) => {
    const { value, e } = event;

    // Open the dropdown if value is truthy
    if (value) {
      openDrop(e);
    }

    // Handle state update depending on whether suggestions is query-controlled
    setState({
      activeSuggestion: 0,
      showSuggestions: true,
      userInput: value
    });

    if (debounced) {
      onChangeDebounced({ value, e });
    } else {
      onChange({ value, e });
    }
  };

  // Handle when the user makes a selection from the dropdown
  const handleSelect = (suggestion) => {
    setState({
      activeSuggestion: 0,
      showSuggestions: false,
      userInput: suggestion.label
    });
    onSelect(suggestion);
  };
  
  return (
    <React.Fragment>
      <Input
        background={ background }
        onChange={ handleChange }
        value={ state.userInput }
        name={ `${ name }-input` }
        { ...inputProps }
      />
      <SuggestionsList
        activeSuggestion={ state.activeSuggestion }
        background={ background }
        dropState={ dropState }
        error={ error }
        errorMessage={ errorMessage }
        handleSelect={ handleSelect }
        loading={ loading }
        retryAction={ retryAction }
        showSuggestions={ state.showSuggestions }
        setDropState={ setDropState }
        suggestions={ suggestions }
        userInput={ state.userInput }
      />
    </React.Fragment>
  );
};

TypeAhead.defaultProps = {
  background: colours.grey,
  debounced: false,
  error: false,
  errorMessage: 'Whoops something went wrong!',
  loading: false,
  onChange: () => null,
  onSelect: () => null,
  retryAction: null,
  suggestions: []
};

TypeAhead.propTypes = {
  background: PropTypes.string,
  debounced: PropTypes.bool,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  loading: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  onSelect: PropTypes.func,
  retryAction: PropTypes.func,
  suggestions: PropTypes.array
};

export default TypeAhead;