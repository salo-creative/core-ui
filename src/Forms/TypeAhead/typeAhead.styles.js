import styled from 'styled-components';

export const BlankWrapper = styled.div`
  height: 4rem;
  display: flex;
  width: 100%;
  position: relative;
  padding: 0 1rem;
  align-items: center;
  font-size: 1.2rem;
  button {
    padding-left: 0;
  }
`;