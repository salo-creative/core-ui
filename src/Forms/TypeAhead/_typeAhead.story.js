import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text } from '@storybook/addon-knobs';
import { State, Store } from '@sambego/storybook-state';
import { filter } from 'lodash';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { TypeAhead } from '../../index';

// README
import README from './README.md';

const stories = storiesOf('Forms | TypeAhead', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['typeahead'] });

stories.add(
  'Basic',
  (() => {
    // STORE
    const store = new Store({ value: '', suggestions: [] });
    // KNOBS
    const label = text('Label', 'TypeAhead label');
    const loading = boolean('Show loading state', false);
    const helperText = text('Helper text', 'Helper text');
    const error = boolean('Show error state', false);
    const required = boolean('Required field', false);
    const disabled = boolean('Show disabled state', false);
    const suggestions = [
      { id: '1', label: 'Sales Adviser' },
      { id: '2', label: 'System Admin' },
      { id: '3', label: 'Organisation Admin' },
      { id: '4', label: 'Marketing' },
      { id: '5', label: 'Client' },
      { id: '6', label: 'Provider' }
    ];

    return (
      <State store={ store }>
        { state => (
          <TypeAhead
            onChange={ ({ value }) => {
              let newSuggestions = [];
              if (value) {
                newSuggestions = filter(suggestions, (suggestion) => {
                  return suggestion.label.toLowerCase().includes(value.toLowerCase());
                });
              }
              store.set({ suggestions: newSuggestions });
            } }
            label={ label }
            loading={ loading }
            helperText={ helperText }
            error={ error }
            required={ required }
            disabled={ disabled }
            suggestions={ state.suggestions }
            name='story'
          />
        ) }
      </State>
    );
  }),
  {
    info: { propTablesExclude: [State] },
    notes: README
  }
);