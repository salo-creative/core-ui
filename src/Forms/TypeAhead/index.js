import TypeAhead from './typeAhead';

export default TypeAhead;