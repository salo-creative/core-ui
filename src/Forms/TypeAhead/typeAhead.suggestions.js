import React from 'react';
import PropTypes from 'prop-types';
import { get, isEmpty } from 'lodash';
import Button from '../../Molecules/Button';
import Drop from '../../Molecules/Drop';
import Loader from '../../Molecules/Loader';

// COMPONENTS & STYLES
import { BlankWrapper } from './typeAhead.styles';

// HELPERS & CONSTANTS
import { colours } from '../../helpers/colours';

const SuggestionsList = (props) => {
  const {
    activeSuggestion,
    background,
    dropState,
    error,
    errorMessage,
    handleSelect,
    loading,
    retryAction,
    setDropState,
    showSuggestions,
    suggestions,
    userInput
  } = props;

  const closeDrop = (e) => {
    if (e.target !== dropState.element && !loading) {
      setDropState({ ...dropState, dropOpen: false });
    }
  };

  const renderItems = () => {
    if (!loading && !error) {
      if (!isEmpty(suggestions)) {
        return suggestions.map((suggestion, index) => (
          <Button
            key={ suggestion.id }
            onClick={ () => handleSelect(suggestion) }
            appearance={ index === activeSuggestion ? 'primary' : 'text' }
            fullWidth
            radius={ false }
            align='start'
            size='M'
          >
            { suggestion.label }
          </Button>
        ));
      }
      return <BlankWrapper>No results found</BlankWrapper>;
    }
    return null;
  };

  const renderLoader = () => {
    if (loading) {
      return (
        <BlankWrapper>
          <Loader
            loaderProps={ {
              size: 40,
              position: 'absolute'
            } }
            display
          />
        </BlankWrapper>
      );
    }
    return null;
  };

  const renderError = () => {
    if (error && !loading) {
      return (
        <BlankWrapper>
          { retryAction && (
            <Button
              onClick={ () => retryAction() }
              iconBefore='sync'
              appearance='text'
            />
          ) }
          <span>{ errorMessage }</span>
        </BlankWrapper>
      );
    }
    return null;
  };

  const renderSuggestions = () => {
    // No input.
    if (!userInput && isEmpty(suggestions)) return null;
    return (
      <Drop
        background={ background }
        width={ get(dropState.element, 'clientWidth', 0) + 2 || 0 }
        disableOverflow={ false }
        element={ dropState.element }
        onClose={ (e) => closeDrop(e) }
        top={ 0 }
        offsetTop={ 2 }
        offsetLeft={ 1 }
        fixed={ false }
        open={ dropState.dropOpen && showSuggestions }
        border={ `1px solid ${ colours.blue }` }
        shadow='none'
        spacing={ 0 }
        zIndex={ 100 }
      >
        { renderLoader() }
        { renderError() }
        { renderItems() }
      </Drop>
    );
  };

  return (
    <React.Fragment>
      { renderSuggestions() }
    </React.Fragment>
  );
};

SuggestionsList.defaultProps = {
  activeSuggestion: 0,
  loading: false,
  retryAction: null,
  showSuggestions: false,
  suggestions: [],
  userInput: ''
};

SuggestionsList.propTypes = {
  activeSuggestion: PropTypes.number,
  background: PropTypes.string.isRequired,
  dropState: PropTypes.object.isRequired,
  error: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string.isRequired,
  retryAction: PropTypes.func,
  setDropState: PropTypes.func.isRequired,
  showSuggestions: PropTypes.bool,
  handleSelect: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  suggestions: PropTypes.array,
  userInput: PropTypes.string
};

export default SuggestionsList;