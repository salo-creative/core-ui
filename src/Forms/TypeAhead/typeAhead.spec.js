import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import TypeAhead from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <TypeAhead />
  );
  expect(container.firstChild).toMatchSnapshot();
});