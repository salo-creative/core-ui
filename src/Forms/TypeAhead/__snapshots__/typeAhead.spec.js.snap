// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 0 2rem;
}

.c1 {
  position: relative;
}

.c2 {
  margin: 0;
  padding: 0 1rem;
  height: 4rem;
  background-color: #e9ecef;
  border: 1px solid;
  border-color: rgba(0,172,237,0.5);
  -webkit-transition: border 0.2s linear;
  transition: border 0.2s linear;
  width: 100%;
  border-radius: 0.4rem;
  font-size: 1.4rem;
}

.c2:focus,
.c2:active {
  border-color: #00ACED;
}

.c2[disabled] {
  cursor: not-allowed;
  opacity: 0.75;
}

<div
  class="c0 salo-input "
>
  <div
    class="c1 salo-input__wrapper"
  >
    <input
      aria-invalid="false"
      aria-required="false"
      class="c2 salo-input__field"
      font-size="1.4rem"
      height="4rem"
      id="undefined-input"
      name="undefined-input"
      placeholder=""
      type="text"
      value=""
    />
  </div>
</div>
`;
